File: Search/RedBlackBST.java
package Search;

import org.omg.PortableInterceptor.SYSTEM_EXCEPTION;

public class RedBlackBST<Key extends Comparable<Key>, Value> {
    private Node root;

    private static final boolean RED = true;
    private static final boolean Black = false;

    private class Node {
        Key key;
        Value val;
        Node left, right;
        int N;
        boolean color;

        Node(Key key, Value val, int N, boolean color) {
            this.key = key;
            this.val = val;
            this.N = N;
            this.color = color;
        }
    }

    public boolean isRed(Node x) {
        if (x == null)
            return false;
        return x.color == RED;
    }

    Node rotateLeft(Node h) {
        Node x = h.right;
        h.right = x.left;
        x.left = h;
        x.color = h.color;
        h.color = RED;
        x.N = h.N;
        h.N = 1 + size(h.left) + size(h.right);
        return x;
    }

    Node rotateRight(Node h) {
        Node x = h.left;
        h.left = x.right;
        x.right = h;
        x.color = h.color;
        h.color = RED;
        x.N = h.N;
        h.N = 1 + size(h.left) + size(h.right);
        return x;
    }

    public int size() {
        return size(root);
    }

    public int size(Node x) {
        if (x == null)
            return 0;
        return x.N;
    }

    public void put(Key key, Value val) {
        root = put(root, key, val);
        root.color = Black;
    }

    void flipColors(Node h) {
        h.color = RED;
        h.left.color = Black;
        h.right.color = Black;
    }

    private Node put(Node h, Key key, Value val) {
        if (h == null)
            return new Node(key, val, 1, RED);
        int cmp = key.compareTo(h.key);
        if (cmp < 0)
            h.left = put(h.left, key, val);
        else if (cmp > 0)
            h.right = put(h.right, key, val);
        else h.val = val;
        if (isRed(h.right) && !isRed(h.left))
            h = rotateLeft(h);
        if (isRed(h.left) && isRed(h.left.left))
            h = rotateRight(h);
        if (isRed(h.left) && isRed(h.right))
            flipColors(h);
        h.N = size(h.left) + size(h.right) + 1;
        return h;
    }

    public void show() {
        Node x = root;
        show(x);
        System.out.println();
    }

    private void show(Node x) {
        if (x == null)
            return;
        System.out.print(x.key);
        show(x.left);
        show(x.right);
    }
}

File: Algorithm_Exercises/Algorithm (4th edition)/Algorithm/Chapter3/RedBlackBST.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Reflection.Metadata.Ecma335;
using System.Text;

namespace Chapter3
{
    public class RedBlackBST<Key,Value>:BST<Key,Value> where Key:IComparable<Key>
    {

        private static readonly bool RED = true;
        private static readonly bool BLACK = false;

        private bool IsRed(RedBlackBSTNode<Key, Value> x) => x != null && x.Color == RED;

        private RedBlackBSTNode<Key, Value> RotateLeft(RedBlackBSTNode<Key, Value> h)
        {
            RedBlackBSTNode<Key, Value> x = (RedBlackBSTNode<Key, Value>)h.Right;
            h.Right = x.Left;
            x.Left = h;
            x.Color = h.Color;
            h.Color = RED;
            x.N = h.N;
            h.N = 1 + GetNodeSize(h.Left) + GetNodeSize(h.Right);
            return x;
        }

        private RedBlackBSTNode<Key, Value> RotateRight(RedBlackBSTNode<Key, Value> h)
        {
            RedBlackBSTNode<Key, Value> x = (RedBlackBSTNode<Key, Value>)h.Left;
            h.Left = x.Right;
            x.Right = h;
            x.Color = h.Color;
            h.Color = RED;
            x.N = h.N;
            h.N = 1 + GetNodeSize(h.Left) + GetNodeSize(h.Right);
            return x;
        }

        public override void Put(Key key, Value value)
        {
            var root = Put((RedBlackBSTNode<Key, Value>)Root, key, value);
            root.Color = BLACK;
            Root = root;
        }

        private RedBlackBSTNode<Key, Value> Put(RedBlackBSTNode<Key, Value> h, Key key, Value value)
        {
            if (h == null) //标准的插入操作，和父节点用红链接相连
                return new RedBlackBSTNode<Key, Value>(key, value, 1, RED);
            int cmp = key.CompareTo(h.Key);
            if (cmp < 0)
                h.Left = Put((RedBlackBSTNode<Key, Value>)h.Left, key, value);
            else if (cmp > 0)
                h.Right = Put((RedBlackBSTNode<Key, Value>)h.Right, key, value);
            else h.Val = value;
            if (IsRed((RedBlackBSTNode<Key, Value>)h.Right) && !IsRed((RedBlackBSTNode<Key, Value>)h.Left)) //左黑右红，左旋转
                h = RotateLeft(h);
            if (IsRed((RedBlackBSTNode<Key, Value>)h.Left) && IsRed((RedBlackBSTNode<Key, Value>)h.Left.Left))//连续两个红，右旋转
                h = RotateRight(h);
            if (IsRed((RedBlackBSTNode<Key, Value>)h.Left) && IsRed((RedBlackBSTNode<Key, Value>)h.Right)) //左右都为红，执行颜色变换
                FlipColours(h);
            h.N = GetNodeSize(h.Left) + GetNodeSize(h.Right) + 1;
            return h;
        }

        private void FlipColours(RedBlackBSTNode<Key, Value> h)
        {
            h.Color = RED;
            ((RedBlackBSTNode<Key, Value>)h.Left).Color = BLACK;
            ((RedBlackBSTNode<Key, Value>)h.Right).Color = BLACK;
        }

        public override void Delete(Key key)
        {
            throw new NotImplementedException();
        }

    }
}
